# **Initialize**

In [1]:
import pandas as pd
import numpy as np
import re
import matplotlib.pyplot as plt
import matplotlib.gridspec as gridspec
from sklearn import preprocessing
import seaborn as sns
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score

%matplotlib inline

**Initialize the data to use**

**Please note that you will need to change the path to the file location**

In [2]:
df = pd.read_csv(r"titanic_data.csv")

In [3]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

**We'll see how big is the data**

In [4]:
df.shape

(891, 12)

**It tells us the data have 891 row**

# **Describing the data**

**With only a function you can know mean, standard deviation,min max, and all sort of things about the data!**

In [5]:
df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

**You can use these functions to know how many unique values does the data have**

In [6]:
df.nunique()

PassengerId    891
Survived         2
Pclass           3
Name           891
Sex              2
Age             88
SibSp            7
Parch            7
Ticket         681
Fare           248
Cabin          147
Embarked         3
dtype: int64

In [55]:
df["Survived"].unique()

array([0, 1])

In [57]:
df["Survived"].value_counts()

0    549
1    342
Name: Survived, dtype: int64

**Check null values**

In [9]:
df.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

# **Working with null values**

**We already know that we need to fill the null values before going further**

**We'll do two operations to fill out the null values by using mean and mode**

### **Mean**

In [10]:
mean_age = df["Age"].mean()
mean_age

29.69911764705882

In [11]:
std_age = df["Age"].std()
std_age

14.526497332334042

In [12]:
is_null_age = df["Age"].isnull().sum()
is_null_age

177

In [13]:
rand_age = np.random.randint(mean_age - std_age, mean_age + std_age, size = is_null_age)
rand_age

array([37, 40, 40, 24, 40, 35, 38, 17, 41, 34, 33, 24, 41, 22, 17, 36, 18,
       31, 35, 18, 29, 31, 43, 41, 40, 18, 36, 36, 27, 38, 31, 40, 37, 28,
       19, 36, 37, 34, 25, 23, 41, 24, 33, 42, 42, 22, 20, 19, 25, 29, 25,
       28, 34, 19, 20, 31, 31, 29, 41, 22, 15, 23, 40, 20, 42, 23, 23, 31,
       33, 30, 39, 35, 31, 38, 18, 26, 20, 30, 30, 28, 29, 33, 25, 18, 24,
       25, 29, 23, 18, 22, 27, 37, 39, 43, 42, 41, 19, 33, 26, 34, 38, 21,
       16, 25, 36, 37, 16, 27, 22, 35, 30, 27, 18, 42, 31, 18, 23, 21, 41,
       19, 39, 23, 17, 43, 43, 28, 29, 37, 29, 39, 28, 32, 17, 38, 43, 40,
       37, 26, 41, 29, 23, 32, 34, 34, 33, 21, 33, 15, 24, 36, 42, 24, 38,
       36, 21, 20, 24, 27, 26, 24, 34, 23, 38, 19, 18, 19, 24, 43, 20, 43,
       27, 31, 41, 20, 40, 30, 20])

In [14]:
age_slice = df.Age.copy()
age_slice[np.isnan(age_slice)] = rand_age
df.Age = age_slice
df.Age = df.Age.astype(int)

df.head(5)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex  Age  SibSp  \
0                            Braund, Mr. Owen Harris    male   22      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female   38      1   
2                             Heikkinen, Miss. Laina  female   26      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female   35      1   
4                           Allen, Mr. William Henry    male   35      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

## **Mode**

In [15]:
df.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age              0
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [16]:
modus = df.Embarked.mode()
modus

0    S
dtype: object

In [17]:
df.Embarked = df.Embarked.fillna(modus.values[0], inplace = False)
df.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age              0
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         0
dtype: int64

In [18]:
deck = {"A": 1, "B": 2, "C": 3, "D": 4, "E": 5, "F": 6, "G": 7, "T": 8, "U": 9}

df["Cabin"] = df["Cabin"].fillna("U0")
df["Deck"] = df["Cabin"].map(lambda x: re.compile("([a-zA-Z]+)").search(x).group())
df["Deck"] = df["Deck"].map(deck)
df["Deck"] = df["Deck"].fillna(0)
df["Deck"] = df["Deck"].astype(int)

In [19]:
df.head(5)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex  Age  SibSp  \
0                            Braund, Mr. Owen Harris    male   22      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female   38      1   
2                             Heikkinen, Miss. Laina  female   26      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female   35      1   
4                           Allen, Mr. William Henry    male   35      0   

   Parch            Ticket     Fare Cabin Embarked  Deck  
0      0         A/5 21171   7.2500    U0        S     9  
1      0          PC 17599  71.2833   C85        C     3  
2      0  STON/O2. 3101282   7.9250    U0        S     9  
3      0            113803  53.1000  C123        S     3  
4      0            373450   8.0500    U0        S     9